File: Packages/EditorKit/Sources/EditorKit/EditorDocument.swift
import CoreGraphics
import Foundation

public struct EditorDocument {
    public let baseImage: CGImage
    public private(set) var annotations: [any Annotation] = []

    public var size: CGSize { CGSize(width: baseImage.width, height: baseImage.height) }

    public init(baseImage: CGImage) { self.baseImage = baseImage }

    public mutating func add(_ a: any Annotation) { annotations.append(a) }

    public mutating func remove(id: UUID) { annotations.removeAll { $0.id == id } }

    public func index(of id: UUID) -> Int? { annotations.firstIndex { $0.id == id } }

    public mutating func move(id: UUID, by delta: CGVector) {
        guard let i = index(of: id) else { return }
        annotations[i] = annotations[i].moved(by: delta)
    }

    public mutating func replace(id: UUID, with a: any Annotation) {
        guard let i = index(of: id) else { return }
        annotations[i] = a
    }

    public mutating func bringToFront(id: UUID) {
        guard let i = index(of: id) else { return }
        let a = annotations.remove(at: i); annotations.append(a)
    }

    public mutating func sendToBack(id: UUID) {
        guard let i = index(of: id) else { return }
        let a = annotations.remove(at: i); annotations.insert(a, at: 0)
    }

    /// Topmost (last-drawn) annotation under the point.
    public func topmostHit(at point: CGPoint) -> UUID? {
        for a in annotations.reversed() where a.hitTest(point) { return a.id }
        return nil
    }

    public func nextCounterNumber() -> Int { 1 } // refined in Task 6
}

File: windows/src/BetterScreenshot.Editor/EditorDocument.cs
using BetterScreenshot.Core;

namespace BetterScreenshot.Editor;

/// <summary>
/// The annotation document: a base-image size plus an ordered list of annotations (array order = z-order,
/// first = back, last = front). The WPF editor holds the actual base bitmap alongside; this pure model carries
/// only the size so it stays headless-testable. All coordinates are base-image pixels, top-left origin.
/// </summary>
public sealed class EditorDocument
{
    private readonly List<IAnnotation> _annotations;

    public PxSize Size { get; }
    public IReadOnlyList<IAnnotation> Annotations => _annotations;

    public EditorDocument(PxSize size, IEnumerable<IAnnotation>? annotations = null)
    {
        Size = size;
        _annotations = annotations?.ToList() ?? new List<IAnnotation>();
    }

    public void Add(IAnnotation a) => _annotations.Add(a);

    public bool Remove(Guid id) => _annotations.RemoveAll(a => a.Id == id) > 0;

    public int? IndexOf(Guid id)
    {
        int i = _annotations.FindIndex(a => a.Id == id);
        return i < 0 ? null : i;
    }

    public void Move(Guid id, double dx, double dy)
    {
        int i = _annotations.FindIndex(a => a.Id == id);
        if (i >= 0) _annotations[i] = _annotations[i].MovedBy(dx, dy);
    }

    public void Replace(Guid id, IAnnotation a)
    {
        int i = _annotations.FindIndex(x => x.Id == id);
        if (i >= 0) _annotations[i] = a;
    }

    public void BringToFront(Guid id)
    {
        int i = _annotations.FindIndex(a => a.Id == id);
        if (i < 0) return;
        var a = _annotations[i];
        _annotations.RemoveAt(i);
        _annotations.Add(a);
    }

    public void SendToBack(Guid id)
    {
        int i = _annotations.FindIndex(a => a.Id == id);
        if (i < 0) return;
        var a = _annotations[i];
        _annotations.RemoveAt(i);
        _annotations.Insert(0, a);
    }

    /// <summary>Topmost (last-drawn) annotation whose lenient hit-test contains the point, or null.</summary>
    public Guid? TopmostHit(PxPoint p)
    {
        for (int i = _annotations.Count - 1; i >= 0; i--)
            if (_annotations[i].HitTest(p)) return _annotations[i].Id;
        return null;
    }

    /// <summary>Ids of annotations whose bounding box intersects <paramref name="rect"/> (marquee selection).</summary>
    public IReadOnlyList<Guid> IdsIntersecting(PxRect rect) =>
        _annotations.Where(a => !a.BoundingBox().Intersection(rect).IsEmpty).Select(a => a.Id).ToList();

    /// <summary>Next counter number = (# of existing counter annotations) + 1.</summary>
    public int NextCounterNumber() => _annotations.OfType<CounterAnnotation>().Count() + 1;

    /// <summary>Crops the document to <paramref name="rect"/>: new size = rect.Size, annotations offset by −rect.origin.</summary>
    public EditorDocument Cropped(PxRect rect)
    {
        var moved = _annotations.Select(a => a.MovedBy(-rect.X, -rect.Y));
        return new EditorDocument(new PxSize(rect.Width, rect.Height), moved);
    }
}
